src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -582,19 +596,102 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+        if (factoryContext.ParameterBinders.Count > 0)
+        {
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, object?[], Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr, BoundValuesArrayExpr).Compile();
+
+            // Looping over arrays is faster
+            var binders = factoryContext.ParameterBinders.ToArray();
+            var count = binders.Length;
+
+            return async (target, httpContext) =>
+            {
+                // Run these first so that they can potentially read and rewind the body
+                var boundValues = new object?[count];
+
+                for (var i = 0; i < count; i++)
+                {
+                    boundValues[i] = await binders[i](httpContext);
+                }
+
+                var (bodyValue, successful) = await TryReadBodyAsync(httpContext);
+
+                if (!successful)
+                {
+                    return;
+                }
+
+                await continuation(target, httpContext, bodyValue, boundValues);
+            };
+        }
         else
         {
-            return HandleRequestBodyAndCompileRequestDelegateForJson(responseWritingMethodCall, factoryContext);
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr).Compile();
+
+            return async (target, httpContext) =>
+            {
+                var (bodyValue, successful) = await TryReadBodyAsync(httpContext);
+
+                if (!successful)
+                {
+                    return;
+                }
+
+                await continuation(target, httpContext, bodyValue);
+            };
+        }
+
+        static async Task<(ReadOnlySequence<byte>, bool)> TryReadBodyAsync(HttpContext httpContext)
+        {
+            var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+
+            if (feature?.CanHaveBody == true)
+            {
+                var bodyReader = httpContext.Request.BodyReader;
+
+                while (true)
+                {
+                    var result = await bodyReader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    if (result.IsCompleted)
+                    {
+                        return (buffer, true);
+                    }
+
+                    // Buffer the body
+                    bodyReader.AdvanceTo(buffer.Start, buffer.End);
+                }
+            }
+
+            return (default, true);
```

Comment:
Did you mean to return false here instead? Otherwise this method only returns true for the flag so isn't really a `Try...()` and could just return the Task<ROS>.

Reply:
👍🏾